src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs

```diff
@@ -0,0 +1,337 @@
+            }
+
+            //Make the calls to the sources
+            foreach (var latestVersionTask in latestVersionsRequests)
+            {
+                contactSourcesRunningTasks.Add(latestVersionTask);
+                //Throttle if needed
+                if (maxTasks <= contactSourcesRunningTasks.Count())
+                {
+                    var finishedTask = await Task.WhenAny(contactSourcesRunningTasks);
+                    contactSourcesRunningTasks.Remove(finishedTask);
+                }
+            }
+            await Task.WhenAll(contactSourcesRunningTasks);
+
+            //Save latest versions within the InstalledPackageReference
+            GetVersionsFromDict(packages, packagesVersionsDict);
+        }
+
+        /// <summary>
+        /// Adding the packages to a unique set to avoid attempting
+        /// to get the latest versions for the same package multiple
+        /// times
+        /// </summary>
+        /// <param name="packages"> Packages found in the project </param>
+        /// <param name="packagesVersionsDict"> An empty dictionary to be filled with packages </param>
+        private void AddPackagesToDict(IEnumerable<FrameworkPackages> packages,
+            Dictionary<InstalledPackageReference, IList<NuGetVersion>> packagesVersionsDict)
+        {
+            foreach (var frameworkPackages in packages)
+            {
+                foreach (var topLevelPackage in frameworkPackages.TopLevelPackages)
+                {
+                    if (!packagesVersionsDict.ContainsKey(topLevelPackage))
+                    {
+                        packagesVersionsDict.Add(topLevelPackage, new List<NuGetVersion>());
+                    }
+                }
+
+                foreach (var transitivePackage in frameworkPackages.TransitivePackages)
+                {
+                    if (!packagesVersionsDict.ContainsKey(transitivePackage))
+                    {
+                        packagesVersionsDict.Add(transitivePackage, new List<NuGetVersion>());
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Get last versions for every package from the unqiue packages
+        /// </summary>
+        /// <param name="packages"> Project packages to get filled with latest versions </param>
+        /// <param name="packagesVersionsDict"> Unique packages that are mapped to latest versions
+        /// from different sources </param>
+        private void GetVersionsFromDict(IEnumerable<FrameworkPackages> packages,
+            Dictionary<InstalledPackageReference, IList<NuGetVersion>> packagesVersionsDict)
+        {
+            foreach (var frameworkPackages in packages)
+            {
+                foreach (var topLevelPackage in frameworkPackages.TopLevelPackages)
+                {
+                    var matchingPackage = packagesVersionsDict.Where(p => p.Key.Name.Equals(topLevelPackage.Name, StringComparison.OrdinalIgnoreCase)).Single();
```

Comment:
instead of `Single` use `First` which is more performance optimized since it won't go through all the items in the list then.